### PR TITLE
C++: Add models for `strlcpy` and `strlcat`

### PR DESCRIPTION
**Pull request by jketema**

### CODE REVIEW

File: cpp/ql/lib/semmle/code/cpp/models/implementations/Strcat.qll
Comment (MathiasVP) [inline cpp/ql/lib/semmle/code/cpp/models/implementations/Strcat.qll:29]: right?
```suggestion
        "strlcat" // strlcat(dst, src, dst_size)
```

**Review by MathiasVP — COMMENTED** [review]
One comment, but otherwise this LGTM (assuming DCA comes back clean!)

File: cpp/ql/lib/semmle/code/cpp/models/implementations/Strcat.qll
Comment (jketema) [reply cpp/ql/lib/semmle/code/cpp/models/implementations/Strcat.qll:29]: Copy paste errors 🎉

**Review by MathiasVP — DISMISSED** [review]

**Comment by jketema** [comment]
Looking at the DCA results, I think these need to be their own separate models. The `cpp/bad-strncpy-size` results look TPs, but it's probably better to deal with that at the query level instead of the model level. All the `cpp/bad-strncpy-size` results look like FPs.

File: cpp/ql/lib/semmle/code/cpp/models/implementations/Strcat.qll
Comment (MathiasVP) [inline cpp/ql/lib/semmle/code/cpp/models/implementations/Strcat.qll:122]: https://linux.die.net/man/3/strlcat says that the return value is "the initial length of dst plus the length of src". I don't think this should have dataflow?

File: cpp/ql/lib/semmle/code/cpp/models/implementations/Strcat.qll
Comment (MathiasVP) [inline cpp/ql/lib/semmle/code/cpp/models/implementations/Strcat.qll:132]: As above: I don't think `output.isReturnValueDeref()` makes sense here as the function apparently returns a `size_t`.

File: cpp/ql/lib/semmle/code/cpp/models/implementations/Strcat.qll
Comment (jketema) [reply cpp/ql/lib/semmle/code/cpp/models/implementations/Strcat.qll:122]: You're right. I had missed that the return value of these functions is totally different.

Comment (MathiasVP) [reply cpp/ql/lib/semmle/code/cpp/models/implementations/Strcat.qll:122]: Yeah, I also missed that initially. Even more reason to split this out into a separate model, I guess 🙂.

Comment (jketema) [reply cpp/ql/lib/semmle/code/cpp/models/implementations/Strcat.qll]: Fixed by https://github.com/github/codeql/pull/14735/commits/5e21a5d28441db320a88cb27bf44589f44a6fb97

File: cpp/ql/lib/semmle/code/cpp/models/implementations/Strcat.qll
Comment (jketema) [reply cpp/ql/lib/semmle/code/cpp/models/implementations/Strcat.qll]: Fixed by https://github.com/github/codeql/pull/14735/commits/5e21a5d28441db320a88cb27bf44589f44a6fb97

File: cpp/ql/lib/semmle/code/cpp/models/implementations/Strcpy.qll
Comment (jketema) [inline cpp/ql/lib/semmle/code/cpp/models/implementations/Strcpy.qll:122]: Probably this should even be:
```suggestion
      or
      not this.returnsTotalLength() and output.isReturnValueDeref()
      or
      this.returnsTotalLength() and output.isReturnValue()
```

Comment (MathiasVP) [reply cpp/ql/lib/semmle/code/cpp/models/implementations/Strcpy.qll:122]: I'm not even sure if we want to taint the return value when it's returning a length. In https://github.com/github/codeql/pull/3592 we removed taint flow from `strlen` because it apparently gave too many FPs, and it may be the case that it's the same thing here 🤔

Comment (jketema) [reply cpp/ql/lib/semmle/code/cpp/models/implementations/Strcpy.qll:122]: In that case is should probably be removed from the `strlcat` model

Comment (MathiasVP) [reply cpp/ql/lib/semmle/code/cpp/models/implementations/Strcpy.qll:122]: Yeah, I think I agree

Comment (jketema) [reply cpp/ql/lib/semmle/code/cpp/models/implementations/Strcpy.qll:122]: Done in https://github.com/github/codeql/pull/14735/commits/1c878750494e87603a8519b5463297cbfb3b6712

File: cpp/ql/lib/semmle/code/cpp/models/implementations/Strcat.qll
Comment (geoffw0) [inline cpp/ql/lib/semmle/code/cpp/models/implementations/Strcat.qll:107]:
```suggestion
  int getParamSize() { result = 2 }
```
As this model is never going to include `strcat`.

File: cpp/ql/lib/semmle/code/cpp/models/implementations/Strcpy.qll
Comment (geoffw0) [inline cpp/ql/lib/semmle/code/cpp/models/implementations/Strcpy.qll:35]:
```suggestion
        "stpncpy", // stpncpy(dest, src, max_amount)
```

**Review by geoffw0 — COMMENTED** [review]
Nit picking...

**Review by MathiasVP — APPROVED** [review]

**Comment by jketema** [comment]
Note that I'm rerunning DCA at the moment just to be sure the latest changes don't break things.

**Comment by jketema** [comment]
DCA looks good to me. The alert changes still look correct.

Does this need a change note?

**Comment by MathiasVP** [comment]
Good point. Yeah, I think we should add a change note 👍

**Comment by jketema** [comment]
Done. Will need re-approval - assuming the change note is ok.

**Review by MathiasVP — APPROVED** [review]
LGTM!

**Pull request merged**
